In [2]:
import pandas as pd

In [3]:
dailyActivity = pd.read_csv("dataset/dailyActivity_merged.csv")
dailyCalories = pd.read_csv("dataset/dailyCalories_merged.csv")
dailyIntensity = pd.read_csv("dataset/dailyIntensities_merged.csv")
dailySteps = pd.read_csv("dataset/dailySteps_merged.csv")
heartrate = pd.read_csv("dataset/heartrate_seconds_merged.csv")
hourCalories = pd.read_csv("dataset/hourlyCalories_merged.csv")
hourIntensity = pd.read_csv("dataset/hourlyIntensities_merged.csv")
hourSteps = pd.read_csv("dataset/hourlySteps_merged.csv")
minCaloriesNarrow = pd.read_csv("dataset/minuteCaloriesNarrow_merged.csv")
minCaloriesWide = pd.read_csv("dataset/minuteCaloriesWide_merged.csv")
minIntensityNarrow = pd.read_csv("dataset/minuteIntensitiesNarrow_merged.csv")
minIntensityWide = pd.read_csv("dataset/minuteIntensitiesWide_merged.csv")
minMetsNarrow = pd.read_csv("dataset/minuteMETsNarrow_merged.csv")
minSleep = pd.read_csv("dataset/minuteSleep_merged.csv")
minStepsNarrow = pd.read_csv("dataset/minuteStepsNarrow_merged.csv")
minStepsWide = pd.read_csv("dataset/minuteStepsWide_merged.csv")
sleepDay = pd.read_csv("dataset/sleepDay_merged.csv")
weightLog = pd.read_csv("dataset/weightLogInfo_merged.csv")

# 1. Procedure

- In order to clean the dataset efficiently and effectively, I will only focus on cleaning the datasets that are essential for my analysis to understand how consumers are using their smart device.
- Datasets to focus: *__dailyActivity_merged.csv__*, *__sleepDay_merged.csv__*, and *__weightLogInfo_merged.csv__*

## 1.1 dailyActivity_merged.csv
- dailyActivity has 940 observations, no missing values, no duplicate rows.
- The daily activity contains records between __12 Apr-12 May 2016__.
- There are 33 distinct users in this dataset and 21 of them use the smart device consecutively for 31 days.
- Converted Id data type and Activity Date data type to 'string' and 'datetime', respectively.
- Added a Day column and TotalMinutes column.

In [3]:
dailyActivity.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        940 non-null    int64  
 1   ActivityDate              940 non-null    object 
 2   TotalSteps                940 non-null    int64  
 3   TotalDistance             940 non-null    float64
 4   TrackerDistance           940 non-null    float64
 5   LoggedActivitiesDistance  940 non-null    float64
 6   VeryActiveDistance        940 non-null    float64
 7   ModeratelyActiveDistance  940 non-null    float64
 8   LightActiveDistance       940 non-null    float64
 9   SedentaryActiveDistance   940 non-null    float64
 10  VeryActiveMinutes         940 non-null    int64  
 11  FairlyActiveMinutes       940 non-null    int64  
 12  LightlyActiveMinutes      940 non-null    int64  
 13  SedentaryMinutes          940 non-null    int64  
 14  Calories  

In [4]:
dailyActivity.isnull().sum()

Id                          0
ActivityDate                0
TotalSteps                  0
TotalDistance               0
TrackerDistance             0
LoggedActivitiesDistance    0
VeryActiveDistance          0
ModeratelyActiveDistance    0
LightActiveDistance         0
SedentaryActiveDistance     0
VeryActiveMinutes           0
FairlyActiveMinutes         0
LightlyActiveMinutes        0
SedentaryMinutes            0
Calories                    0
dtype: int64

In [5]:
# Check duplicates
dailyActivity.duplicated().sum()

0

In [6]:
# Check duplicates on Id and ActivityDate for any duplicate entries per Id
dailyActivity.duplicated(subset=['Id','ActivityDate']).sum()

0

In [7]:
dailyActivity

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0    1503960366    4/12/2016       13162       8.500000         8.500000   
1    1503960366    4/13/2016       10735       6.970000         6.970000   
2    1503960366    4/14/2016       10460       6.740000         6.740000   
3    1503960366    4/15/2016        9762       6.280000         6.280000   
4    1503960366    4/16/2016       12669       8.160000         8.160000   
..          ...          ...         ...            ...              ...   
935  8877689391     5/8/2016       10686       8.110000         8.110000   
936  8877689391     5/9/2016       20226      18.250000        18.250000   
937  8877689391    5/10/2016       10733       8.150000         8.150000   
938  8877689391    5/11/2016       21420      19.559999        19.559999   
939  8877689391    5/12/2016        8064       6.120000         6.120000   

     LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                         0.0                1.88                      0.55   
1                         0.0                1.57                      0.69   
2                         0.0                2.44                      0.40   
3                         0.0                2.14                      1.26   
4                         0.0                2.71                      0.41   
..                        ...                 ...                       ...   
935                       0.0                1.08                      0.20   
936                       0.0               11.10                      0.80   
937                       0.0                1.35                      0.46   
938                       0.0               13.22                      0.41   
939                       0.0                1.82                      0.04   

     LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                   6.06                     0.00                 25   
1                   4.71                     0.00                 21   
2                   3.91                     0.00                 30   
3                   2.83                     0.00                 29   
4                   5.04                     0.00                 36   
..                   ...                      ...                ...   
935                 6.80                     0.00                 17   
936                 6.24                     0.05                 73   
937                 6.28                     0.00                 18   
938                 5.89                     0.00                 88   
939                 4.25                     0.00                 23   

     FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                     13                   328               728      1985  
1                     19                   217               776      1797  
2                     11                   181              1218      1776  
3                     34                   209               726      1745  
4                     10                   221               773      1863  
..                   ...                   ...               ...       ...  
935                    4                   245              1174      2847  
936                   19                   217              1131      3710  
937                   11                   224              1187      2832  
938                   12                   213              1127      3832  
939                    1                   137               770      1849  

[940 rows x 15 columns]

In [8]:
# Find earliest date of record
dailyActivity.sort_values(by='ActivityDate', ascending=True, inplace=False)

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0    1503960366    4/12/2016       13162           8.50             8.50   
799  8253242879    4/12/2016        9033           7.16             7.16   
412  4388161847    4/12/2016       10122           7.78             7.78   
381  4319703577    4/12/2016        7753           5.20             5.20   
377  4057192912    4/12/2016        5394           4.03             4.03   
..          ...          ...         ...            ...              ...   
408  4319703577     5/9/2016       10378           6.96             6.96   
439  4388161847     5/9/2016       10218           7.86             7.86   
470  4445114986     5/9/2016        5275           3.54             3.54   
907  8792009665     5/9/2016           0           0.00             0.00   
343  3977333714     5/9/2016       14687          10.08            10.08   

     LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                         0.0                1.88                      0.55   
799                       0.0                5.43                      0.14   
412                       0.0                0.00                      0.00   
381                       0.0                0.00                      0.00   
377                       0.0                0.00                      0.00   
..                        ...                 ...                       ...   
408                       0.0                0.14                      0.56   
439                       0.0                0.34                      0.73   
470                       0.0                0.00                      0.00   
907                       0.0                0.00                      0.00   
343                       0.0                0.77                      5.60   

     LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                   6.06                      0.0                 25   
799                 1.59                      0.0                 40   
412                 0.00                      0.0                  0   
381                 0.00                      0.0                  0   
377                 3.94                      0.0                  0   
..                   ...                      ...                ...   
408                 6.25                      0.0                  2   
439                 6.79                      0.0                  6   
470                 3.54                      0.0                  0   
907                 0.00                      0.0                  0   
343                 3.55                      0.0                  8   

     FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                     13                   328               728      1985  
799                    2                   154              1244      2044  
412                    0                     0              1440      2955  
381                    0                     0              1440      2115  
377                    0                   164              1276      2286  
..                   ...                   ...               ...       ...  
408                   14                   374               490      2345  
439                   19                   258              1020      3013  
470                    0                   266               641      2281  
907                    0                     0              1440      1688  
343                  122                   151              1159      1667  

[940 rows x 15 columns]

In [9]:
# Find latest date of record
dailyActivity.sort_values(by='ActivityDate', ascending=False, inplace=False)

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
470  4445114986     5/9/2016        5275           3.54             3.54   
795  8053475328     5/9/2016       12465           9.38             9.38   
439  4388161847     5/9/2016       10218           7.86             7.86   
27   1503960366     5/9/2016       12022           7.72             7.72   
181  2022484408     5/9/2016       13379           9.39             9.39   
..          ...          ...         ...            ...              ...   
377  4057192912    4/12/2016        5394           4.03             4.03   
216  2320127002    4/12/2016       10725           7.49             7.49   
880  8792009665    4/12/2016        2564           1.64             1.64   
185  2026352035    4/12/2016        4414           2.74             2.74   
0    1503960366    4/12/2016       13162           8.50             8.50   

     LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
470                       0.0                0.00                      0.00   
795                       0.0                6.12                      0.57   
439                       0.0                0.34                      0.73   
27                        0.0                3.45                      0.53   
181                       0.0                2.12                      1.63   
..                        ...                 ...                       ...   
377                       0.0                0.00                      0.00   
216                       0.0                1.17                      0.31   
880                       0.0                0.00                      0.00   
185                       0.0                0.19                      0.35   
0                         0.0                1.88                      0.55   

     LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
470                 3.54                      0.0                  0   
795                 2.69                      0.0                 66   
439                 6.79                      0.0                  6   
27                  3.74                      0.0                 46   
181                 5.64                      0.0                 35   
..                   ...                      ...                ...   
377                 3.94                      0.0                  0   
216                 6.01                      0.0                 13   
880                 1.64                      0.0                  0   
185                 2.20                      0.0                  3   
0                   6.06                      0.0                 25   

     FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
470                    0                   266               641      2281  
795                   12                   148              1214      2765  
439                   19                   258              1020      3013  
27                    11                   206               835      1819  
181                   47                   297              1061      2709  
..                   ...                   ...               ...       ...  
377                    0                   164              1276      2286  
216                    9                   306              1112      2124  
880                    0                   116               831      2044  
185                    8                   181               706      1459  
0                     13                   328               728      1985  

[940 rows x 15 columns]

In [10]:
# Check distinct Id columns
dailyActivity['Id'].value_counts()

1503960366    31
4319703577    31
8583815059    31
8378563200    31
8053475328    31
7086361926    31
6962181067    31
5553957443    31
4702921684    31
4558609924    31
1624580081    31
4388161847    31
4445114986    31
8877689391    31
1927972279    31
2873212765    31
2320127002    31
4020332650    31
2026352035    31
1844505072    31
2022484408    31
3977333714    30
1644430081    30
5577150313    30
8792009665    29
6290855005    29
6117666160    28
6775888955    26
7007744171    26
3372868164    20
8253242879    19
2347167796    18
4057192912     4
Name: Id, dtype: int64

In [11]:
filter_user = dailyActivity[dailyActivity['Id'] == 1503960366]
filter_user

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0   1503960366    4/12/2016       13162           8.50             8.50   
1   1503960366    4/13/2016       10735           6.97             6.97   
2   1503960366    4/14/2016       10460           6.74             6.74   
3   1503960366    4/15/2016        9762           6.28             6.28   
4   1503960366    4/16/2016       12669           8.16             8.16   
5   1503960366    4/17/2016        9705           6.48             6.48   
6   1503960366    4/18/2016       13019           8.59             8.59   
7   1503960366    4/19/2016       15506           9.88             9.88   
8   1503960366    4/20/2016       10544           6.68             6.68   
9   1503960366    4/21/2016        9819           6.34             6.34   
10  1503960366    4/22/2016       12764           8.13             8.13   
11  1503960366    4/23/2016       14371           9.04             9.04   
12  1503960366    4/24/2016       10039           6.41             6.41   
13  1503960366    4/25/2016       15355           9.80             9.80   
14  1503960366    4/26/2016       13755           8.79             8.79   
15  1503960366    4/27/2016       18134          12.21            12.21   
16  1503960366    4/28/2016       13154           8.53             8.53   
17  1503960366    4/29/2016       11181           7.15             7.15   
18  1503960366    4/30/2016       14673           9.25             9.25   
19  1503960366     5/1/2016       10602           6.81             6.81   
20  1503960366     5/2/2016       14727           9.71             9.71   
21  1503960366     5/3/2016       15103           9.66             9.66   
22  1503960366     5/4/2016       11100           7.15             7.15   
23  1503960366     5/5/2016       14070           8.90             8.90   
24  1503960366     5/6/2016       12159           8.03             8.03   
25  1503960366     5/7/2016       11992           7.71             7.71   
26  1503960366     5/8/2016       10060           6.58             6.58   
27  1503960366     5/9/2016       12022           7.72             7.72   
28  1503960366    5/10/2016       12207           7.77             7.77   
29  1503960366    5/11/2016       12770           8.13             8.13   
30  1503960366    5/12/2016           0           0.00             0.00   

    LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                        0.0                1.88                      0.55   
1                        0.0                1.57                      0.69   
2                        0.0                2.44                      0.40   
3                        0.0                2.14                      1.26   
4                        0.0                2.71                      0.41   
5                        0.0                3.19                      0.78   
6                        0.0                3.25                      0.64   
7                        0.0                3.53                      1.32   
8                        0.0                1.96                      0.48   
9                        0.0                1.34                      0.35   
10                       0.0                4.76                      1.12   
11                       0.0                2.81                      0.87   
12                       0.0                2.92                      0.21   
13                       0.0                5.29                      0.57   
14                       0.0                2.33                      0.92   
15                       0.0                6.40                      0.41   
16                       0.0                3.54                      1.16   
17                       0.0                1.06                      0.50   
18                       0.0                3.56                      1.42   
19                       0.0                2.29   

In [12]:
# Convert Id data type from 'int64' to 'string' and ActivityDate data type from 'string' to 'datetime'
dailyActivity['Id'] = dailyActivity['Id'].astype('string')
dailyActivity['ActivityDate'] = pd.to_datetime(dailyActivity['ActivityDate'])
dailyActivity.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Id                        940 non-null    string        
 1   ActivityDate              940 non-null    datetime64[ns]
 2   TotalSteps                940 non-null    int64         
 3   TotalDistance             940 non-null    float64       
 4   TrackerDistance           940 non-null    float64       
 5   LoggedActivitiesDistance  940 non-null    float64       
 6   VeryActiveDistance        940 non-null    float64       
 7   ModeratelyActiveDistance  940 non-null    float64       
 8   LightActiveDistance       940 non-null    float64       
 9   SedentaryActiveDistance   940 non-null    float64       
 10  VeryActiveMinutes         940 non-null    int64         
 11  FairlyActiveMinutes       940 non-null    int64         
 12  LightlyActiveMinutes  

In [13]:
dailyActivity['ActivityDate'].dt.day

0      12
1      13
2      14
3      15
4      16
       ..
935     8
936     9
937    10
938    11
939    12
Name: ActivityDate, Length: 940, dtype: int64

In [14]:
filter_user = dailyActivity[dailyActivity['Id'] == '1503960366']
filter_user

Id ActivityDate  TotalSteps  TotalDistance  TrackerDistance  \
0   1503960366   2016-04-12       13162           8.50             8.50   
1   1503960366   2016-04-13       10735           6.97             6.97   
2   1503960366   2016-04-14       10460           6.74             6.74   
3   1503960366   2016-04-15        9762           6.28             6.28   
4   1503960366   2016-04-16       12669           8.16             8.16   
5   1503960366   2016-04-17        9705           6.48             6.48   
6   1503960366   2016-04-18       13019           8.59             8.59   
7   1503960366   2016-04-19       15506           9.88             9.88   
8   1503960366   2016-04-20       10544           6.68             6.68   
9   1503960366   2016-04-21        9819           6.34             6.34   
10  1503960366   2016-04-22       12764           8.13             8.13   
11  1503960366   2016-04-23       14371           9.04             9.04   
12  1503960366   2016-04-24       10039           6.41             6.41   
13  1503960366   2016-04-25       15355           9.80             9.80   
14  1503960366   2016-04-26       13755           8.79             8.79   
15  1503960366   2016-04-27       18134          12.21            12.21   
16  1503960366   2016-04-28       13154           8.53             8.53   
17  1503960366   2016-04-29       11181           7.15             7.15   
18  1503960366   2016-04-30       14673           9.25             9.25   
19  1503960366   2016-05-01       10602           6.81             6.81   
20  1503960366   2016-05-02       14727           9.71             9.71   
21  1503960366   2016-05-03       15103           9.66             9.66   
22  1503960366   2016-05-04       11100           7.15             7.15   
23  1503960366   2016-05-05       14070           8.90             8.90   
24  1503960366   2016-05-06       12159           8.03             8.03   
25  1503960366   2016-05-07       11992           7.71             7.71   
26  1503960366   2016-05-08       10060           6.58             6.58   
27  1503960366   2016-05-09       12022           7.72             7.72   
28  1503960366   2016-05-10       12207           7.77             7.77   
29  1503960366   2016-05-11       12770           8.13             8.13   
30  1503960366   2016-05-12           0           0.00             0.00   

    LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                        0.0                1.88                      0.55   
1                        0.0                1.57                      0.69   
2                        0.0                2.44                      0.40   
3                        0.0                2.14                      1.26   
4                        0.0                2.71                      0.41   
5                        0.0                3.19                      0.78   
6                        0.0                3.25                      0.64   
7                        0.0                3.53                      1.32   
8                        0.0                1.96                      0.48   
9                        0.0                1.34                      0.35   
10                       0.0                4.76                      1.12   
11                       0.0                2.81                      0.87   
12                       0.0                2.92                      0.21   
13                       0.0                5.29                      0.57   
14                       0.0                2.33                      0.92   
15                       0.0                6.40                      0.41   
16                       0.0                3.54                      1.16   
17                       0.0                1.06                      0.50   
18                       0.0                3.56                      1.42   
19                       0.0                2.29   

In [15]:
# Add Day column and TotalMinutes column
dailyActivity['Day'] = dailyActivity['ActivityDate'].dt.day_name()
dailyActivity['TotalMinutes'] = dailyActivity['VeryActiveMinutes'] + dailyActivity['FairlyActiveMinutes'] + dailyActivity['LightlyActiveMinutes'] + dailyActivity['SedentaryMinutes']

In [16]:
dailyActivity.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 17 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   Id                        940 non-null    string        
 1   ActivityDate              940 non-null    datetime64[ns]
 2   TotalSteps                940 non-null    int64         
 3   TotalDistance             940 non-null    float64       
 4   TrackerDistance           940 non-null    float64       
 5   LoggedActivitiesDistance  940 non-null    float64       
 6   VeryActiveDistance        940 non-null    float64       
 7   ModeratelyActiveDistance  940 non-null    float64       
 8   LightActiveDistance       940 non-null    float64       
 9   SedentaryActiveDistance   940 non-null    float64       
 10  VeryActiveMinutes         940 non-null    int64         
 11  FairlyActiveMinutes       940 non-null    int64         
 12  LightlyActiveMinutes  

In [17]:
# Export the cleaned dailyActivity dataset
dailyActivity.to_csv("dataset/dailyActivity_clean.csv", index=False)

## 1.2 sleepDay_merged.csv
- sleepDay has 413 observations, no missing values, and 3 duplicate rows.
- The sleep day was recorded between __12 Apr-12 May 2016__.
- There are 24 distinct users in this dataset and 3 of them use the smart device during sleep consecutively for 31 days.
- Converted Id data type and SleepDay data type to 'string' and 'datetime', respectively.
- Added a Day column.

In [18]:
sleepDay.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 413 entries, 0 to 412
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Id                  413 non-null    int64 
 1   SleepDay            413 non-null    object
 2   TotalSleepRecords   413 non-null    int64 
 3   TotalMinutesAsleep  413 non-null    int64 
 4   TotalTimeInBed      413 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 16.3+ KB


In [19]:
# Convert Id data type to string and SleepDay data type to datetime
sleepDay['Id'] = sleepDay['Id'].astype('string')
sleepDay['SleepDay'] = pd.to_datetime(sleepDay['SleepDay'])
sleepDay.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 413 entries, 0 to 412
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   Id                  413 non-null    string        
 1   SleepDay            413 non-null    datetime64[ns]
 2   TotalSleepRecords   413 non-null    int64         
 3   TotalMinutesAsleep  413 non-null    int64         
 4   TotalTimeInBed      413 non-null    int64         
dtypes: datetime64[ns](1), int64(3), string(1)
memory usage: 16.3 KB


In [20]:
sleepDay

Id   SleepDay  TotalSleepRecords  TotalMinutesAsleep  \
0    1503960366 2016-04-12                  1                 327   
1    1503960366 2016-04-13                  2                 384   
2    1503960366 2016-04-15                  1                 412   
3    1503960366 2016-04-16                  2                 340   
4    1503960366 2016-04-17                  1                 700   
..          ...        ...                ...                 ...   
408  8792009665 2016-04-30                  1                 343   
409  8792009665 2016-05-01                  1                 503   
410  8792009665 2016-05-02                  1                 415   
411  8792009665 2016-05-03                  1                 516   
412  8792009665 2016-05-04                  1                 439   

     TotalTimeInBed  
0               346  
1               407  
2               442  
3               367  
4               712  
..              ...  
408             360  
409             527  
410             423  
411             545  
412             463  

[413 rows x 5 columns]

In [21]:
sleepDay.isnull().sum()

Id                    0
SleepDay              0
TotalSleepRecords     0
TotalMinutesAsleep    0
TotalTimeInBed        0
dtype: int64

In [22]:
# Check duplicates
sleepDay.duplicated(keep=False).sum()

6

In [23]:
sleepDay[sleepDay.duplicated(keep=False)]

Id   SleepDay  TotalSleepRecords  TotalMinutesAsleep  \
160  4388161847 2016-05-05                  1                 471   
161  4388161847 2016-05-05                  1                 471   
222  4702921684 2016-05-07                  1                 520   
223  4702921684 2016-05-07                  1                 520   
379  8378563200 2016-04-25                  1                 388   
380  8378563200 2016-04-25                  1                 388   

     TotalTimeInBed  
160             495  
161             495  
222             543  
223             543  
379             402  
380             402

In [24]:
# Remove duplicates
sleepDay.drop_duplicates(inplace=True, ignore_index=True)
sleepDay.duplicated(keep=False).sum()

0

In [25]:
# Find earliest date of record
sleepDay.sort_values(by='SleepDay', ascending=True, inplace=False)

Id   SleepDay  TotalSleepRecords  TotalMinutesAsleep  \
0    1503960366 2016-04-12                  1                 327   
109  4020332650 2016-04-12                  1                 501   
166  4445114986 2016-04-12                  2                 429   
199  4702921684 2016-04-12                  1                 425   
226  5553957443 2016-04-12                  1                 441   
..          ...        ...                ...                 ...   
360  7086361926 2016-05-12                  1                 444   
64   2026352035 2016-05-12                  1                 456   
394  8378563200 2016-05-12                  1                 496   
334  6962181067 2016-05-12                  1                 516   
256  5553957443 2016-05-12                  1                 438   

     TotalTimeInBed  
0               346  
109             541  
166             457  
199             439  
226             464  
..              ...  
360             457  
64              485  
394             546  
334             535  
256             475  

[410 rows x 5 columns]

In [26]:
# Find latest date of record
sleepDay.sort_values(by='SleepDay', ascending=False, inplace=False)

Id   SleepDay  TotalSleepRecords  TotalMinutesAsleep  \
193  4445114986 2016-05-12                  1                 483   
334  6962181067 2016-05-12                  1                 516   
256  5553957443 2016-05-12                  1                 438   
360  7086361926 2016-05-12                  1                 444   
394  8378563200 2016-05-12                  1                 496   
..          ...        ...                ...                 ...   
226  5553957443 2016-04-12                  1                 441   
257  5577150313 2016-04-12                  1                 419   
337  7086361926 2016-04-12                  1                 514   
304  6962181067 2016-04-12                  1                 366   
0    1503960366 2016-04-12                  1                 327   

     TotalTimeInBed  
193             501  
334             535  
256             475  
360             457  
394             546  
..              ...  
226             464  
257             438  
337             525  
304             387  
0               346  

[410 rows x 5 columns]

In [27]:
sleepDay['Id'].value_counts()

8378563200    31
6962181067    31
5553957443    31
2026352035    28
3977333714    28
4445114986    28
4702921684    27
5577150313    26
4319703577    26
1503960366    25
7086361926    24
4388161847    23
6117666160    18
8792009665    15
2347167796    15
4020332650     8
1927972279     5
4558609924     5
1644430081     4
6775888955     3
8053475328     3
1844505072     3
7007744171     2
2320127002     1
Name: Id, dtype: Int64

In [28]:
# Add Day column
sleepDay['Day'] = sleepDay['SleepDay'].dt.day_name()

In [29]:
# Export the cleaned sleepDay dataset
sleepDay.to_csv("dataset/sleepDay_clean.csv", index=False)

## 1.3 weightLogInfo_merged.csv
- weightLogInfo has 67 observations, 65 missing values for the *__Fat__* column, no duplicate rows.
- The weight log info contains records between __12 Apr-12 May 2016__.
- There are 8 distinct users in this dataset and only 1 of them use the smart device to record weight for 30 days.
- Converted Id and LogId data type and Date data type to 'string' and 'datetime', respectively.
- Added a Day column.
- Do not understand the meaning of IsManualReport. 
- A LogId is possible to be used by 3 different users.

In [30]:
weightLog.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67 entries, 0 to 66
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Id              67 non-null     int64  
 1   Date            67 non-null     object 
 2   WeightKg        67 non-null     float64
 3   WeightPounds    67 non-null     float64
 4   Fat             2 non-null      float64
 5   BMI             67 non-null     float64
 6   IsManualReport  67 non-null     bool   
 7   LogId           67 non-null     int64  
dtypes: bool(1), float64(4), int64(2), object(1)
memory usage: 3.9+ KB


In [31]:
# Convert Id and LogId data type to string and Date data type to datetime
weightLog['Id'] = weightLog['Id'].astype('string')
weightLog['LogId'] = weightLog['LogId'].astype('string')
weightLog['Date'] = pd.to_datetime(weightLog['Date'])

In [32]:
weightLog.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67 entries, 0 to 66
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Id              67 non-null     string        
 1   Date            67 non-null     datetime64[ns]
 2   WeightKg        67 non-null     float64       
 3   WeightPounds    67 non-null     float64       
 4   Fat             2 non-null      float64       
 5   BMI             67 non-null     float64       
 6   IsManualReport  67 non-null     bool          
 7   LogId           67 non-null     string        
dtypes: bool(1), datetime64[ns](1), float64(4), string(2)
memory usage: 3.9 KB


In [33]:
weightLog.isnull().sum()

Id                 0
Date               0
WeightKg           0
WeightPounds       0
Fat               65
BMI                0
IsManualReport     0
LogId              0
dtype: int64

In [34]:
weightLog.duplicated().sum()

0

In [35]:
weightLog['Id'].value_counts()

6962181067    30
8877689391    24
4558609924     5
1503960366     2
2873212765     2
4319703577     2
1927972279     1
5577150313     1
Name: Id, dtype: Int64

In [36]:
# Find the earliest record
weightLog.sort_values(by='Date', ascending=True, inplace=False)

Id                Date    WeightKg  WeightPounds  Fat        BMI  \
43  8877689391 2016-04-12 06:47:11   85.800003    189.156628  NaN  25.680000   
13  6962181067 2016-04-12 23:59:59   62.500000    137.788914  NaN  24.389999   
2   1927972279 2016-04-13 01:08:52  133.500000    294.317120  NaN  47.540001   
44  8877689391 2016-04-13 06:55:00   84.900002    187.172464  NaN  25.410000   
14  6962181067 2016-04-13 23:59:59   62.099998    136.907061  NaN  24.240000   
..         ...                 ...         ...           ...  ...        ...   
65  8877689391 2016-05-11 06:51:47   85.400002    188.274775  NaN  25.559999   
41  6962181067 2016-05-11 23:59:59   61.900002    136.466144  NaN  24.170000   
66  8877689391 2016-05-12 06:42:53   84.000000    185.188300  NaN  25.139999   
42  6962181067 2016-05-12 23:59:59   61.900002    136.466144  NaN  24.170000   
4   2873212765 2016-05-12 23:59:59   57.299999    126.324875  NaN  21.690001   

    IsManualReport          LogId  
43           False  1460443631000  
13            True  1460505599000  
2            False  1460509732000  
44           False  1460530500000  
14            True  1460591999000  
..             ...            ...  
65           False  1462949507000  
41            True  1463011199000  
66           False  1463035373000  
42            True  1463097599000  
4             True  1463097599000  

[67 rows x 8 columns]

In [37]:
# Find the latest record
weightLog.sort_values(by='Date', ascending=False, inplace=False)

Id                Date    WeightKg  WeightPounds  Fat        BMI  \
42  6962181067 2016-05-12 23:59:59   61.900002    136.466144  NaN  24.170000   
4   2873212765 2016-05-12 23:59:59   57.299999    126.324875  NaN  21.690001   
66  8877689391 2016-05-12 06:42:53   84.000000    185.188300  NaN  25.139999   
41  6962181067 2016-05-11 23:59:59   61.900002    136.466144  NaN  24.170000   
65  8877689391 2016-05-11 06:51:47   85.400002    188.274775  NaN  25.559999   
..         ...                 ...         ...           ...  ...        ...   
14  6962181067 2016-04-13 23:59:59   62.099998    136.907061  NaN  24.240000   
44  8877689391 2016-04-13 06:55:00   84.900002    187.172464  NaN  25.410000   
2   1927972279 2016-04-13 01:08:52  133.500000    294.317120  NaN  47.540001   
13  6962181067 2016-04-12 23:59:59   62.500000    137.788914  NaN  24.389999   
43  8877689391 2016-04-12 06:47:11   85.800003    189.156628  NaN  25.680000   

    IsManualReport          LogId  
42            True  1463097599000  
4             True  1463097599000  
66           False  1463035373000  
41            True  1463011199000  
65           False  1462949507000  
..             ...            ...  
14            True  1460591999000  
44           False  1460530500000  
2            False  1460509732000  
13            True  1460505599000  
43           False  1460443631000  

[67 rows x 8 columns]

In [38]:
# Keep only the date, remove the time
weightLog['Date'] = weightLog['Date'].dt.date
weightLog

Id        Date    WeightKg  WeightPounds   Fat        BMI  \
0   1503960366  2016-05-02   52.599998    115.963147  22.0  22.650000   
1   1503960366  2016-05-03   52.599998    115.963147   NaN  22.650000   
2   1927972279  2016-04-13  133.500000    294.317120   NaN  47.540001   
3   2873212765  2016-04-21   56.700001    125.002104   NaN  21.450001   
4   2873212765  2016-05-12   57.299999    126.324875   NaN  21.690001   
..         ...         ...         ...           ...   ...        ...   
62  8877689391  2016-05-06   85.000000    187.392923   NaN  25.440001   
63  8877689391  2016-05-08   85.400002    188.274775   NaN  25.559999   
64  8877689391  2016-05-09   85.500000    188.495234   NaN  25.610001   
65  8877689391  2016-05-11   85.400002    188.274775   NaN  25.559999   
66  8877689391  2016-05-12   84.000000    185.188300   NaN  25.139999   

    IsManualReport          LogId  
0             True  1462233599000  
1             True  1462319999000  
2            False  1460509732000  
3             True  1461283199000  
4             True  1463097599000  
..             ...            ...  
62           False  1462517015000  
63           False  1462692953000  
64           False  1462775984000  
65           False  1462949507000  
66           False  1463035373000  

[67 rows x 8 columns]

In [39]:
print(len(weightLog['LogId'].value_counts()))
weightLog['LogId'].value_counts()

56


1462233599000    3
1462319999000    2
1462147199000    2
1461628799000    2
1461023999000    2
1462838399000    2
1460937599000    2
1463097599000    2
1461283199000    2
1462406399000    2
1460678399000    1
1461110399000    1
1460851199000    1
1460764799000    1
1460509732000    1
1460591999000    1
1460505599000    1
1460884675000    1
1461455999000    1
1461196799000    1
1461887999000    1
1463011199000    1
1461542399000    1
1461801599000    1
1460443631000    1
1461974399000    1
1462060799000    1
1462492799000    1
1462579199000    1
1462751999000    1
1462924799000    1
1461369599000    1
1462692953000    1
1462665599000    1
1461653427000    1
1461739865000    1
1460962274000    1
1460813965000    1
1461047971000    1
1461134694000    1
1461221427000    1
1461396148000    1
1461483485000    1
1461566416000    1
1462092469000    1
1460616523000    1
1461826203000    1
1461912595000    1
1462002543000    1
1460530500000    1
1462258181000    1
1462344502000    1
146251701500

In [40]:
weightLog[weightLog['LogId'] == '1462233599000']

Id        Date   WeightKg  WeightPounds   Fat        BMI  \
0   1503960366  2016-05-02  52.599998    115.963147  22.0  22.650000   
10  4558609924  2016-05-02  69.199997    152.559879   NaN  27.040001   
32  6962181067  2016-05-02  61.500000    135.584291   NaN  24.000000   

    IsManualReport          LogId  
0             True  1462233599000  
10            True  1462233599000  
32            True  1462233599000

In [42]:
# Add Day column
weightLog['Date'] = pd.to_datetime(weightLog['Date'])
weightLog['Day'] = weightLog['Date'].dt.day_name()

In [43]:
# Export the cleaned weightLog dataset
weightLog.to_csv("dataset/weightLog_clean.csv", index=False)

## 1.4 minuteSleep_merged.csv
- minuteSleep has 188521 observations, no missing values, 543 duplicate rows.
- The weight log info contains records between __11 Apr-09 May 2016__.
- There are 24 distinct users in this dataset.
- Converted date data type to 'datetime'.
- Added day column.
- Value 1 represents deep sleep, value 2 represents light sleep, value 3 represents rapid eye movement sleep. 

In [10]:
minSleep.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187978 entries, 0 to 187977
Data columns (total 4 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   Id      187978 non-null  int64 
 1   date    187978 non-null  object
 2   value   187978 non-null  int64 
 3   logId   187978 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 5.7+ MB


In [11]:
minSleep['Id'].value_counts()

5553957443    15682
8378563200    15064
2026352035    15054
6962181067    14450
4319703577    13051
3977333714    12912
5577150313    11976
4702921684    11832
4445114986    11671
7086361926    11194
4388161847     9734
1503960366     9580
6117666160     9183
2347167796     7370
8792009665     6807
4020332650     3038
1844505072     2883
1927972279     2189
1644430081     1384
6775888955     1107
8053475328      905
4558609924      700
7007744171      143
2320127002       69
Name: Id, dtype: int64

In [18]:
minSleep

Id                  date  value        logId
0       1503960366  4/12/2016 2:47:30 AM      3  11380564589
1       1503960366  4/12/2016 2:48:30 AM      2  11380564589
2       1503960366  4/12/2016 2:49:30 AM      1  11380564589
3       1503960366  4/12/2016 2:50:30 AM      1  11380564589
4       1503960366  4/12/2016 2:51:30 AM      1  11380564589
...            ...                   ...    ...          ...
188516  8792009665   5/4/2016 9:59:00 AM      1  11552534115
188517  8792009665  5/4/2016 10:00:00 AM      1  11552534115
188518  8792009665  5/4/2016 10:01:00 AM      1  11552534115
188519  8792009665  5/4/2016 10:02:00 AM      1  11552534115
188520  8792009665  5/4/2016 10:03:00 AM      1  11552534115

[188521 rows x 4 columns]

In [19]:
minSleep.duplicated().sum()

543

In [20]:
minSleep[minSleep.duplicated(keep=False)]

Id                 date  value        logId
99227   4702921684  5/6/2016 9:10:00 PM      3  11573168523
99228   4702921684  5/6/2016 9:11:00 PM      3  11573168523
99229   4702921684  5/6/2016 9:12:00 PM      2  11573168523
99230   4702921684  5/6/2016 9:13:00 PM      1  11573168523
99231   4702921684  5/6/2016 9:14:00 PM      1  11573168523
...            ...                  ...    ...          ...
100308  4702921684  5/7/2016 6:08:00 AM      2  11573168523
100309  4702921684  5/7/2016 6:09:00 AM      1  11573168523
100310  4702921684  5/7/2016 6:10:00 AM      2  11573168523
100311  4702921684  5/7/2016 6:11:00 AM      1  11573168523
100312  4702921684  5/7/2016 6:12:00 AM      1  11573168523

[1086 rows x 4 columns]

In [21]:
minSleep[minSleep.duplicated()]

Id                 date  value        logId
99770   4702921684  5/6/2016 9:10:00 PM      3  11573168523
99771   4702921684  5/6/2016 9:11:00 PM      3  11573168523
99772   4702921684  5/6/2016 9:12:00 PM      2  11573168523
99773   4702921684  5/6/2016 9:13:00 PM      1  11573168523
99774   4702921684  5/6/2016 9:14:00 PM      1  11573168523
...            ...                  ...    ...          ...
100308  4702921684  5/7/2016 6:08:00 AM      2  11573168523
100309  4702921684  5/7/2016 6:09:00 AM      1  11573168523
100310  4702921684  5/7/2016 6:10:00 AM      2  11573168523
100311  4702921684  5/7/2016 6:11:00 AM      1  11573168523
100312  4702921684  5/7/2016 6:12:00 AM      1  11573168523

[543 rows x 4 columns]

In [4]:
# Remove duplicates
minSleep.drop_duplicates(inplace=True, ignore_index=True)
minSleep.duplicated(keep=False).sum()

0

In [5]:
minSleep

Id                  date  value        logId
0       1503960366  4/12/2016 2:47:30 AM      3  11380564589
1       1503960366  4/12/2016 2:48:30 AM      2  11380564589
2       1503960366  4/12/2016 2:49:30 AM      1  11380564589
3       1503960366  4/12/2016 2:50:30 AM      1  11380564589
4       1503960366  4/12/2016 2:51:30 AM      1  11380564589
...            ...                   ...    ...          ...
187973  8792009665   5/4/2016 9:59:00 AM      1  11552534115
187974  8792009665  5/4/2016 10:00:00 AM      1  11552534115
187975  8792009665  5/4/2016 10:01:00 AM      1  11552534115
187976  8792009665  5/4/2016 10:02:00 AM      1  11552534115
187977  8792009665  5/4/2016 10:03:00 AM      1  11552534115

[187978 rows x 4 columns]

In [6]:
minSleep['value'].value_counts()

1    171960
2     14002
3      2016
Name: value, dtype: int64

In [7]:
minSleep.sort_values(by=['Id', 'date'], ascending=True, inplace=False)

Id                  date  value        logId
0       1503960366  4/12/2016 2:47:30 AM      3  11380564589
1       1503960366  4/12/2016 2:48:30 AM      2  11380564589
2       1503960366  4/12/2016 2:49:30 AM      1  11380564589
3       1503960366  4/12/2016 2:50:30 AM      1  11380564589
4       1503960366  4/12/2016 2:51:30 AM      1  11380564589
...            ...                   ...    ...          ...
187969  8792009665   5/4/2016 9:55:00 AM      2  11552534115
187970  8792009665   5/4/2016 9:56:00 AM      1  11552534115
187971  8792009665   5/4/2016 9:57:00 AM      1  11552534115
187972  8792009665   5/4/2016 9:58:00 AM      1  11552534115
187973  8792009665   5/4/2016 9:59:00 AM      1  11552534115

[187978 rows x 4 columns]

In [9]:
minSleep.sort_values(by='date', ascending=False, inplace=False)

Id                   date  value        logId
179874  8378563200    5/9/2016 9:59:30 PM      1  11595151537
29677   2026352035    5/9/2016 9:59:00 PM      1  11602073203
128192  5577150313    5/9/2016 9:59:00 PM      1  11600721447
179873  8378563200    5/9/2016 9:58:30 PM      1  11595151537
128191  5577150313    5/9/2016 9:58:00 PM      1  11600721447
...            ...                    ...    ...          ...
13852   1927972279  4/11/2016 10:01:00 PM      1  11372414035
16109   2026352035  4/11/2016 10:01:00 PM      1  11372566564
13851   1927972279  4/11/2016 10:00:00 PM      1  11372414035
16108   2026352035  4/11/2016 10:00:00 PM      2  11372566564
51484   4020332650  4/11/2016 10:00:00 PM      1  11375838887

[187978 rows x 4 columns]

In [26]:
# Add day column
minSleep['date'] = pd.to_datetime(minSleep['date'])
minSleep['day'] = minSleep['date'].dt.day_name()
minSleep.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 187978 entries, 0 to 187977
Data columns (total 5 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   Id      187978 non-null  int64         
 1   date    187978 non-null  datetime64[ns]
 2   value   187978 non-null  int64         
 3   logId   187978 non-null  int64         
 4   day     187978 non-null  object        
dtypes: datetime64[ns](1), int64(3), object(1)
memory usage: 7.2+ MB


In [27]:
minSleep.head()

Id                date  value        logId      day
0  1503960366 2016-04-12 02:47:30      3  11380564589  Tuesday
1  1503960366 2016-04-12 02:48:30      2  11380564589  Tuesday
2  1503960366 2016-04-12 02:49:30      1  11380564589  Tuesday
3  1503960366 2016-04-12 02:50:30      1  11380564589  Tuesday
4  1503960366 2016-04-12 02:51:30      1  11380564589  Tuesday

In [28]:
# Export the cleaned weightLog dataset
minSleep.to_csv("dataset/minSleep_clean.csv", index=False)